packages/flutter/lib/src/gestures/pointer_data_resampler.dart

```diff
@@ -0,0 +1,199 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:ui' as ui;
+
+/// Class for pointer data resampling.
+class PointerDataResampler {
+  // Data enqueued for processing.
+  final _queuedData = <ui.PointerData>[];
+
+  /// Pointer state required for resampling.
+  ui.PointerData _last;
+  ui.PointerData _next;
+  var _position = ui.Offset(0.0, 0.0);
+  var _isTracked = false;
+  var _isDown = false;
+  var _pointerIdentifier = 0;
+
+  ui.PointerData _clone(
+      ui.PointerData data,
+      ui.PointerChange change,
+      ui.Offset position,
+      ui.Offset delta,
+      int pointerIdentifier,
+      Duration timeStamp) {
+    return ui.PointerData(
+        buttons: data.buttons,
+        device: data.device,
+        timeStamp: timeStamp,
+        change: change,
+        kind: data.kind,
+        physicalDeltaX: delta.dx,
+        physicalDeltaY: delta.dy,
+        physicalX: position.dx,
+        physicalY: position.dy,
+        pointerIdentifier: pointerIdentifier,
+        synthesized: data.synthesized);
+  }
+
+  ui.Offset _positionAt(Duration sampleTime) {
+    // Use `next` position by default.
+    var x = _next.physicalX;
+    var y = _next.physicalY;
+
+    // Resample if `next` time stamp is past `sampleTime`.
+    if (_next.timeStamp > sampleTime && _next.timeStamp > _last.timeStamp) {
+      final interval =
+          (_next.timeStamp - _last.timeStamp).inMicroseconds.toDouble();
+      final scalar =
+          (sampleTime - _last.timeStamp).inMicroseconds.toDouble() / interval;
+      x = _last.physicalX + (_next.physicalX - _last.physicalX) * scalar;
+      y = _last.physicalY + (_next.physicalY - _last.physicalY) * scalar;
+    }
+
+    return ui.Offset(x, y);
+  }
+
+  void _processPointerData(Duration sampleTime) {
+    for (var i = 0; i < _queuedData.length; i++) {
+      final data = _queuedData[i];
+
+      // Update both `last` and `next` pointer data if time stamp is older
+      // or equal to `sampleTime`.
+      if (data.timeStamp <= sampleTime || _last == null) {
+        _last = data;
+        _next = data;
+        continue;
+      }
+
+      // Update only `next` pointer data if time stamp is more recent than
+      // `sampleTime` and next data is not already more recent.
+      if (_next.timeStamp < sampleTime) {
+        _next = data;
+      }
+    }
+  }
+
+  List<ui.PointerData> _dequeueAndSamplePointerDataUntil(Duration sampleTime) {
+    final dequeuedData = <ui.PointerData>[];
+
+    while (_queuedData.isNotEmpty) {
+      final data = _queuedData.first;
+
+      // Potentially stop dequeuing data if more recent than `sampleTime`.
+      if (data.timeStamp > sampleTime) {
+        // Stop if change is not `up` or `remove`, which are allowed to be
+        // processed early as this improves resampling of these changes,
+        // which is important for fling animations.
```

Comment:
Nit: `Stop if change is not up or remove. Otherwise, continue to allow early processing of up and remove events as ...` seems to be clearer than `Stop if change is not up or remove, which are allowed to be processed early as ...` to me.

Reply:
Done